File: uikit-sample/src/main/java/com/sendbird/uikit_messaging_android/widgets/WaitingDialog.java
package com.sendbird.uikit_messaging_android.widgets;

import android.app.Dialog;
import android.content.Context;
import android.content.DialogInterface;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.sendbird.uikit_messaging_android.R;

public class WaitingDialog {
    private final static String TAG = "logger";
    private final static Handler mainHandler = new Handler(Looper.getMainLooper());
    private static final Object waitingDialogLock = new Object();
    private static Dialog waitingDialog;


    private static Dialog getWaitingDialog(@NonNull Context context) {
        synchronized (waitingDialogLock) {
            if (waitingDialog != null) {
                return waitingDialog;
            }

            waitingDialog = new Dialog(context, R.style.Widget_Sendbird_SendbirdProgressDialog);
            return waitingDialog;
        }
    }

    public static void show(@NonNull Context context) {
        show(context, false);
    }

    public static void show(@NonNull Context context, int layoutResId) {
        show(context, false, layoutResId, null);
    }

    public static void show(@NonNull Context context, final boolean cancelable) {
        show(context, cancelable, 0, null);
    }

    public static void show(@NonNull Context context, final boolean cancelable, final int layoutResId, @Nullable DialogInterface.OnCancelListener listener) {
        dismiss();

        mainHandler.post(() -> {
            Log.d(TAG, ">> WaitingDialog::show()");
            waitingDialog = getWaitingDialog(context);
            // here we set layout of progress dialog
            if (layoutResId <= 0) {
                waitingDialog.setContentView(R.layout.sb_view_waiting_dialog);
            } else {
                waitingDialog.setContentView(layoutResId);
            }
            waitingDialog.setCancelable(cancelable);
            if (listener != null) {
                waitingDialog.setOnCancelListener(listener);
            }
            waitingDialog.show();
        });
    }

    public static void dismiss() {
        mainHandler.post(() -> {
            try {
                Log.d(TAG, ">> WaitingDialog::cancel()");
                if (waitingDialog != null) {
                    synchronized (waitingDialogLock) {
                        waitingDialog.cancel();
                        waitingDialog = null;
                    }
                }
            } catch (Exception e) {
                Log.d(TAG, "", e);
            }
        });
    }
}

File: uikit-samples/src/main/java/com/sendbird/uikit/samples/common/widgets/WaitingDialog.kt
package com.sendbird.uikit.samples.common.widgets

import android.app.Dialog
import android.content.Context
import android.content.DialogInterface
import android.os.Handler
import android.os.Looper
import android.util.Log
import com.sendbird.uikit.samples.R

object WaitingDialog {
    private const val TAG = "logger"
    private val mainHandler = Handler(Looper.getMainLooper())
    private val waitingDialogLock = Any()
    private var waitingDialog: Dialog? = null
    private fun getWaitingDialog(context: Context): Dialog {
        synchronized(waitingDialogLock) {
            if (waitingDialog != null) {
                return requireNotNull(waitingDialog)
            }
            waitingDialog = Dialog(context, R.style.Widget_Sendbird_SendbirdProgressDialog)
            return requireNotNull(waitingDialog)
        }
    }

    fun show(context: Context, layoutResId: Int) {
        show(context, false, layoutResId, null)
    }

    @JvmOverloads
    fun show(
        context: Context,
        cancelable: Boolean = false,
        layoutResId: Int = 0,
        listener: DialogInterface.OnCancelListener? = null
    ) {
        dismiss()
        mainHandler.post {
            Log.d(TAG, ">> WaitingDialog::show()")
            val waitingDialog = getWaitingDialog(context)
            // here we set layout of progress dialog
            if (layoutResId <= 0) {
                waitingDialog.setContentView(R.layout.sb_view_waiting_dialog)
            } else {
                waitingDialog.setContentView(layoutResId)
            }
            waitingDialog.setCancelable(cancelable)
            if (listener != null) {
                waitingDialog.setOnCancelListener(listener)
            }
            waitingDialog.show()
        }
    }

    fun dismiss() {
        mainHandler.post {
            try {
                Log.d(TAG, ">> WaitingDialog::cancel()")
                synchronized(waitingDialogLock) {
                    waitingDialog?.cancel()
                    waitingDialog = null
                }
            } catch (e: Exception) {
                Log.d(TAG, "", e)
            }
        }
    }
}
